src/sentry/types/region.py

```diff
@@ -1,8 +1,120 @@
+    maximum bit length specified by our snowflake ID schema.
+    """
+
+    address: str
+    """The address of the region's silo.
+
+    In a production environment, this would typically be a subdomain (e.g.,
+    "eu.sentry.io"). In a development environment, it may be useful to define
+    region addresses as a custom hostname or port number (e.g,
+    "localhost:8001").
+    """
+
+    category: RegionCategory
+    """The region's category."""
+
+    def __post_init__(self) -> None:
+        from sentry.utils.snowflake import REGION_ID
+
+        REGION_ID.validate(self.id)
+
+    def to_url(self, path: str) -> str:
+        return self.address + path
+
+
+class RegionMapping:
+    """The set of all regions in this Sentry platform instance."""
+
+    def __init__(self, regions: Iterable[Region]) -> None:
+        self.regions = tuple(regions)
+        self._by_name = {r.name: r for r in self.regions}
+        self._by_id = {r.id: r for r in self.regions}
+
+    @classmethod
+    def load_from_config(cls) -> RegionMapping:
+        from django.conf import settings
+
+        # For now, assume that all region configs can be taken in through Django
+        # settings. We may investigate other ways of delivering those configs in
+        # production.
+        return cls(settings.SENTRY_REGION_CONFIG)
+
+    def get_by_name(self, name: str) -> Region | None:
+        """Look up a region by name."""
+        return self._by_name.get(name)
+
+    def get_by_id(self, id: int) -> Region | None:
+        """Look up a region by numeric ID."""
+        return self._by_id.get(id)
+
+    def get_for_organization(self, organization: Organization) -> Region | None:
+        """Resolve an organization to the region where its data is stored.
+
+        Returns None if this Sentry platform instance is configured to run only
+        in monolith mode.
+        """
+
+        if not self.regions:
+            return None
+
+        # Backend representation to be determined. If you are working on code
+        # that depends on this method, you can mock it out in unit tests or
+        # temporarily hard-code a placeholder.
+        raise NotImplementedError
+
+    def get_local_region(self) -> Region | None:
+        """Get the region in which this server instance is running.
+
+        Returns None if this server instance is not a region silo.
+        """
+
+        from django.conf import settings
+
+        if SiloMode.get_current_mode() != SiloMode.REGION:
+            return None
+
+        local_region_name = settings.SENTRY_REGION
+        if not local_region_name:
+            raise Exception("SENTRY_REGION must be set when server is in REGION silo mode")
+        local_region = self.get_by_name(local_region_name)
+        if not local_region:
+            raise Exception(f"SENTRY_REGION value not resolved to region: {local_region_name!r}")
+        return local_region
+
+
+_global_region_mapping = None
+
+
+def get_region_mapping() -> RegionMapping:
```

Comment:
This feels ugly. I'd prefer to do something like our common "registry" pattern where the region mapping is just initialized with the module, but that leads to circular imports because `sentry/conf.py` needs to import the `Region` and `RegionMapping` types.

Reply:
a slightly less ugly version of this is to use `functools.lru_cache(maxsize=1)` to hide the `global` part

Reply:
I agree, actually getting the region with `get_region_mapping().get_local_region()` is a bit wordy. Maybe if this is a class and we shorten to `RegionMap`?
e.g.
```python
class RegionMap:
    def __new__(self):
        return (
            RegionMapping.load_from_config()
            if not _global_region_mapping
            else _global_region_mapping
        )
```
but it's essentially the same thing: `RegionMap().get_local_region()`

Reply:
What about an api like,

```
get_local_region() -> str
get_all_regions() -> List[str]
get_region(key: str) -> Region
```

? I don't think we need the Mapping object directly.

Reply:
IE: Essentially flatten the mapping object into module level functions and don't expose it directly.

Reply:
Having a public API made of functions feels better than a global API of shared global data. I recognize that the functions are just wrapping the global state. But the functions *could* be cloning data so that the global state can't be mutated.

Reply:
Module-level interfaces make my Java brain sad but I see how it's nicer to abstract away the mapping object. 😅

Part of the reason I didn't like that originally, was that `override_settings` wouldn't play nice with having the regions cached. (I don't believe the caching is premature optimization, because the mapping could otherwise produce O(n<sup>2</sup>) costs once we have a growing number of single-tenant regions.) In addition to injecting the region values, you would have to mess around with mocks in the `sentry.types.region` module, which usually would be two steps removed from whatever you're trying to test, [as is the case in `test_snowflake`](https://github.com/getsentry/sentry/blob/33f63667de66578f0e014556a69a0d3dec5056a2/tests/sentry/utils/test_snowflake.py#L68).

My solution is [a helper named `override_region` in `testutils`](https://github.com/getsentry/sentry/blob/ceafdf59d6daa5d663ed2d1949f1a340f4e5be2c/src/sentry/testutils/region.py), which hides away the module-level mocking for reuse in other tests. I think [the current `test_snowflake` implementation](https://github.com/getsentry/sentry/blob/ceafdf59d6daa5d663ed2d1949f1a340f4e5be2c/tests/sentry/utils/test_snowflake.py#L69-L73) works out pretty nicely, where it only has to inject `SENTRY_REGION` but doesn't have to directly mock any functions.

Reply:
Is it also possible to "clear" this cached value via a pytest after hook that can be installed globally?

Reply:
Not without using a more sophisticated cache than `functools.lru_cache`, as far as I know.